src/mouse_events.cpp

```diff
@@ -1476,16 +1507,58 @@ std::set<map_location> mouse_handler::get_adj_enemies(const map_location& loc, i
 void mouse_handler::show_attack_options(const unit_map::const_iterator& u)
 {
 	// Cannot attack if no attacks are left.
-	if(u->attacks_left() == 0) {
+	if(u->attacks_left() == 0 || u->attacks().empty()) {
 		return;
 	}
 
 	// Get the teams involved.
 	const team& cur_team = current_team();
 	const team& u_team = pc_.get_teams()[u->side() - 1];
 
-	// Check each adjacent hex.
-	for(const map_location& loc : get_adjacent_tiles(u->get_location())) {
+
+	// convert to cubical -> enumerate tiles -> convert back
+	// see https://www.redblobgames.com/grids/hexagons/ for detailed description
+
+
+	// convert to cubical coordinates
+	// cx == a.x, cy is not needed
+	const auto& attacks = u->attacks();
+	std::set<int> attackable_distances;
+
+    for (const auto& attack : attacks) {
+        for (int i = attack.min_range(); i <= attack.max_range(); ++i) {
+            attackable_distances.insert(i);
+        }
+    }
+
+	// for optimalization, I think iterating over ranges here, would be better
+
+	int min = *attackable_distances.begin();
+	int n = *std::prev(attackable_distances.end());
```

Comment:
Why are these dereferences allowed? I suggest checking for empty.

Reply:
What do you mean?

Reply:
I fo check if the attacks are empty already

Reply:
The set is known to be non-empty because the only way for it to be empty is if there are no attacks, and if there are no attacks then it doesn't even reach this line due to the check on line 1510.

Both dereferences are guaranteed to be valid if there's at least one element, and we know there's at least one element. (I suppose you could add `assert(!attackable_distances.empty());` above these lines to make it clearer that we know there's at least one element.)

Reply:
Ok

Reply:
Unless there is code elsewhere that checks min and max range there is no reason to expect the set to be non-empty. I suggest to not add an assert.

Reply:
I added assert, if the set is empty already.
You've seen that?

Reply:
Yes, hence now it will abort the process instead of more reasonable behavior.

Reply:
It won't reach that assert.

Reply:
No clue how you come to that conclusion. Where is the code that guarantees it?

Reply:
> No clue how you come to that conclusion. Where is the code that guarantees it?

Tell me example, it could fail.

Reply:
Pick min and max ranges so nothing is added to the set.

Reply:
Isn't that true only for invalid min and max ranges?

Reply:
Can you elaborate on what that is and where that is handled?

Reply:
> Can you elaborate on what that is and where that is handled?

I'm sure it happen only when min_range > max_range,
Which is invalid. ...
But I see you point, I'll add a fallback if someone sets those numbers wrongly.

Reply:
*instead of the assert.

Reply:
(Same for the othor UI PR)